crypto/sig_abstractions.go

```diff
@@ -0,0 +1,166 @@
+// Copyright (C) 2019-2021 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package crypto
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/algorand/go-algorand/protocol"
+)
+
+type (
+	//ByteSignature using unspecified bound.
+	//msgp:allocbound ByteSignature
+	ByteSignature []byte
+
+	// AlgorithmType enum type for signing algorithms
+	AlgorithmType uint64
+)
+
+// all AlgorithmType enums
+const (
+	minAlgorithmType AlgorithmType = iota
+
+	DilithiumType
+	Ed25519Type
+
+	maxAlgorithmType
+)
+
+// Signer interface represents the possible operations that can be done with a signing key.
+type Signer interface {
+	Sign(message Hashable) ByteSignature
+	SignBytes(message []byte) ByteSignature
+	GetVerifyingKey() *VerifyingKey
+}
+
+// ErrBadSignature represents a bad signature
+var ErrBadSignature = fmt.Errorf("invalid signature")
+
+// Verifier interface represents any algorithm that can verify signatures for a specific signing scheme.
+type Verifier interface {
+	// Verify and VerifyBytes returns error on bad signature, and any other problem.
+	Verify(message Hashable, sig ByteSignature) error
+	VerifyBytes(message []byte, sig ByteSignature) error
+}
+
+// SignatureAlgorithm holds a Signer, and the type of algorithm the Signer conforms with.
+// to add a key - verify that PackedSignatureAlgorithm's function (getSigner) returns your key.
+type SignatureAlgorithm struct {
```

Comment:
I think `SignatureAlgorithm` is not the best name, since it sounds like it represents an algorithm, not a signer + verifier for a specific algorithm. Perhaps `GenericSigningKey` would be better?

Then maybe `VerifyingKey` could be `GenericVerifyingKey`.